clang/lib/CodeGen/TargetInfo.cpp

```diff
@@ -9015,14 +9015,26 @@ void AMDGPUTargetCodeGenInfo::setTargetAttributes(
     GV->setDSOLocal(true);
   }
 
-  if (GV->isDeclaration())
-    return;
+  // Move the GV->isDeclaration() check till after properly setting
+  // device function attributes.
+
   const FunctionDecl *FD = dyn_cast_or_null<FunctionDecl>(D);
   if (!FD)
     return;
 
   llvm::Function *F = cast<llvm::Function>(GV);
 
+  if (M.getLangOpts().OpenMP) {
```

Comment:
This sort of toggling O0/noinline/alwaysinline might be related to the function pointer hashing, in which case we can drop it in favour of fixed-function-abi for trunk

Reply:
What is fixed-function-abi?